include/socket.h

```diff
@@ -0,0 +1,134 @@
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef SOCKET_H
+# define SOCKET_H
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdint.h>
+#include "network_interface.h"
+
+/** Returned by a secure tcp socket when a transaction is close */
+#define TLS_RET_CLOSE_NOTIFY	-0x7880
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct socket_desc
+ * @brief Socket descriptor
+ */
+struct socket_desc;
+
+/**
+ * @struct tcp_socket_init_param
+ * @brief Parameter to initialize a TCP Socket
+ */
+struct tcp_socket_init_param {
+	//No parameters
+};
+
+/**
+ * @struct stcp_socket_init_param
+ * @brief Parameter to initialize a TCP Socket
+ */
+struct stcp_socket_init_param {
+	char				*ca_cert;
+	char				*hostname;
+	char				*client_cert;
+	char				*client_key;
+};
+
+enum socket_type {
+	TCP_SOCKET,
+	SECURE_TCP_SOCKET
+};
+
+/**
+ * @struct socket_init_param
+ * @brief Parameter to initialize a Socket
+ */
+struct socket_init_param {
+	/** Network interface. Ex: Wifi, Ethernet */
+	struct network_interface	*net;
+	/** Reference to one of the values */
+	enum socket_type		type;
+	/** Socket init param */
+	union {
+		struct tcp_socket_init_param	*tcp;
```

Comment:
this one does not have any members, so a union like this looks a wierd.

Reply:
```
        //socket.c
 	switch (param->type) {
	case TCP_SOCKET:
		init_param = (void *)param->init_param.tcp;
		break;
	case SECURE_TCP_SOCKET:
		init_param = (void *)param->init_param.secure_tcp;
		break;
}
```
This is how it is used. With the type you choose which init param to use.
For the type TCP_SOCKET no init param is needed. I created it just for consistency.
For example if new a type is added (ex. UDP_SOCKET), new param should be added too.